pkg/issuer/acme/dns/util/wait.go

```diff
@@ -66,32 +70,42 @@ func getNameservers(path string, defaults []string) []string {
 	return systemNameservers
 }
 
-// Update FQDN with CNAME if any
-func updateDomainWithCName(r *dns.Msg, fqdn string) string {
-	for _, rr := range r.Answer {
-		if cn, ok := rr.(*dns.CNAME); ok {
-			if cn.Hdr.Name == fqdn {
-				logf.V(logf.DebugLevel).Infof("Updating FQDN: %s with its CNAME: %s", fqdn, cn.Target)
-				fqdn = cn.Target
-				break
+// Update FQDN with CNAME if any, it will follow CNAME records till it hits a non-CNAME.
+// this will error if there is a recursive CNAME in the chain
+func updateDomainWithCName(fqdn string, nameservers []string, fqdnChain ...string) (string, error) {
+	r, err := dnsQuery(fqdn, dns.TypeCNAME, nameservers, true)
+	if err == nil && r.Rcode == dns.RcodeSuccess {
+		for _, rr := range r.Answer {
+			if cn, ok := rr.(*dns.CNAME); ok {
+				if cn.Hdr.Name == fqdn {
+					logf.V(logf.DebugLevel).Infof("Updating FQDN: %s with its CNAME: %s", fqdn, cn.Target)
+
+					// check if we were here before to prevent recursive records causing issues
+					for _, fqdnInChain := range fqdnChain {
+						if cn.Target == fqdnInChain {
+							return "", fmt.Errorf("Found recursive CNAME record to %q when looking up %q", cn.Target, fqdn)
```

Comment:
[RFC 1034](https://tools.ietf.org/html/rfc1034#section-3.6.2) seems to use the term "CNAME loop" instead of "recursive CNAME", should we use that instead?

Reply:
```suggestion
							return "", fmt.Errorf("Found CNAME loop on %q when looking up %q", cn.Target, fqdn)
```